### PR TITLE
nixos/tests/phosh: init

### PR DESCRIPTION
**Pull request by zhaofengli**
###### Description of changes

Makes Phosh updates easier to test.

<details>
<summary>Screenshots in <code>$out</code></summary>

01lockinfo
02unlock
03launcher
04settings
05keyboard
</details>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zhaofengli** [comment]
@ofborg test phosh

**Review by tomfitzhenry — APPROVED** [review]
LGTM with a few minor comments.

File: nixos/tests/phosh.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    import time
+
+    start_all()
+    phone.wait_for_unit("phosh.service")
+
+    with subtest("Check that we can see the lock screen info page"):
+        # Saturday, January 1
+        phone.succeed("timedatectl set-time '2022-01-01 07:00'")
+
+        phone.wait_for_text("Saturday")
```
[the remark this answers is not in the document]
Comment (zhaofengli) [reply nixos/tests/phosh.nix:45]: `timedatectl set-time` sets the local time (UTC in the VM), and the timezone of the test VM doesn't depend on that of the host machine.

File: nixos/tests/phosh.nix [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        phone.screenshot("01lockinfo")
+
+    with subtest("Check that we can unlock the screen"):
+        phone.send_chars("${pin}", delay=0.2)
```
[the remark this answers is not in the document]
Comment (zhaofengli) [reply nixos/tests/phosh.nix:49]: Sure thing. I'm not entirely sure what the cause of the jankiness is, but I suspect it has something to do with the animations.

**Comment by tfc** [comment]
@zhaofengli LGTM. can you please rebase?

**Comment by zhaofengli** [comment]
Rebased and tested with #183913. It worked with the Phoc update in #183907, but with the Phosh update applied it crashed with "No schemas installed".

**Comment by tfc** [comment]
@ofborg test phosh

**Comment by zhaofengli** [comment]
The test doesn't work now before #183913 is merged.

**Comment by tfc** [comment]
ok didn't get that before. So let's get that merged in first.

**Comment by zhaofengli** [comment]
@ofborg test phosh

**Comment by tfc** [comment]
It seems like the test ran through on arm but didn't build on x86. can you please have a look @zhaofengli ?

**Comment by zhaofengli** [comment]
Hmm, looking at the logs there seem to be a huge number of rebuilds which I didn't encounter locally, and it failed building an unrelated package:

```
25/29 test-shortcut-theme               OK              0.24s
26/29 test-model-filter                 OK              0.08s
27/29 test-fuzzy-highlight              OK              0.05s
28/29 test-mutable-scoring              OK              0.05s
29/29 test-recursive-monitor            OK              0.15s

Summary of Failures:

24/29 test-suggestion-buffer    FAIL            0.27s   killed by signal 5 SIGTRAP


Ok:                 28  
Expected Fail:      0   
Fail:               1   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   

Full log written to /build/libdazzle-3.44.0/build/meson-logs/testlog.txt
TEST    vsynth2-huffyuvbgr24
TEST    vsynth2-huffyuvbgra
error: builder for '/nix/store/2kydag4y4j1qk1la663rdvsvg0jywiy9-libdazzle-3.44.0.drv' failed with exit code 1;
```

**Comment by tfc** [comment]
https://github.com/ofborg test phosh

**Comment by tfc** [comment]
> This is passing:
> 
>     * on x86_64 ofborg: https://github.com/NixOS/nixpkgs/pull/173108/checks?check_run_id=7687825361
> 
>     * locally (tested via `nix-build . -A nixosTests.phosh`).
> 
> 
> This is failing on aarch64 ofborg: https://logs.nix.ci/?key=nixos/nixpkgs.173108&attempt_id=455cc027-e508-480b-9d56-786702b17dd7
> Screenshots in `$out`

Ok, this time the ofborg error is related to this actual test... as long as this is supposed to work on aarch64, we gonna get this test fixed, regardless if it works on some other machine. In the future it will has to be solid in the CI, too. Any chance that there might be some big-noisy-machine-multicore-flakiness?

File: nixos/tests/phosh.nix [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        phocConfig = {
+          outputs.Virtual-1 = {
+            scale = 2;
+          };
+        };
+      };
+
+      systemd.services.phosh = {
+        environment = {
+          # Accelerated graphics fail on phoc 0.20 (wlroots 0.15)
```
[the remark this answers is not in the document]
Comment (zhaofengli) [reply nixos/tests/phosh.nix:30]: Yeah, I was testing with #183907 and without using the software renderer it failed saying "Could not find needed globals" with `outputs` being NULL:

```
phone # [  118.480569] systemd[838]: Starting Phosh, a shell for mobile phones...
phone # [  118.746481] .phosh-wrapped[4390]: Could not find needed globals
phone # outputs: 0, layer_shell: 0x118d100, idle_manager: 0x1196510, inhibit: 0x1196490, xdg_wm: 0x118d000, xdg_output: 0x118d080, wlr_output_manager: 0x1142400, wlr_foreign_toplevel_manager: 0x114f740[  118.816159] traps: .phosh-wrapped[4390] trap int3 ip:4a044f sp:7fffda8b42c0 error:0 in .phosh-wrapped[419000+c1000]
```

**Review by tomfitzhenry — APPROVED** [review]
LGTM. Tests are passing on @ofborg.

**Review by Mindavi — APPROVED** [review]

**Comment by Mindavi** [comment]
I like this! Great job adding this :)

File: nixos/tests/phosh.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (jtojnar) [reply nixos/tests/phosh.nix:56]: What happens if the `delay` kwarg is not used?

[the remark this answers is not in the document]
Comment (Mindavi) [reply nixos/tests/phosh.nix:56]: Seems to me like it defaults to 0.01s by default

**Pull request merged**